libs/core/langchain_core/messages/tool.py

```diff
@@ -1,26 +1,80 @@
-from typing import Any, List, Literal
+from typing import Any, List, Literal, Optional
 
+from langchain_core.load import Serializable
+from langchain_core.messages.ai import AIMessage
 from langchain_core.messages.base import (
     BaseMessage,
     BaseMessageChunk,
     merge_content,
 )
 from langchain_core.utils._merge import merge_dicts
+from langchain_core.utils.tools import parse_tool_calls
 
 
-class ToolMessage(BaseMessage):
-    """Message for passing the result of executing a tool back to a model."""
+class ToolCall(Serializable):
+    name: str
+    args: dict
+    id: Optional[str] = None
 
-    tool_call_id: str
-    """Tool call that this message is responding to."""
 
-    type: Literal["tool"] = "tool"
+class ToolCallsMessage(AIMessage):
+    tool_calls: Optional[List[ToolCall]] = None
+
+
+class ToolCallsMessageChunk(ToolCallsMessage, BaseMessageChunk):
+    # Ignoring mypy re-assignment here since we're overriding the value
+    # to make sure that the chunk variant can be discriminated from the
+    # non-chunk variant.
+    type: Literal["ToolCallsMessageChunk"] = "ToolCallsMessageChunk"  # type: ignore[assignment] # noqa: E501
 
     @classmethod
     def get_lc_namespace(cls) -> List[str]:
         """Get the namespace of the langchain object."""
         return ["langchain", "schema", "messages"]
 
+    def __add__(self, other: Any) -> BaseMessageChunk:  # type: ignore
+        if isinstance(other, ToolCallsMessageChunk):
+            if self.example != other.example:
+                raise ValueError(
+                    "Cannot concatenate ToolCallsMessageChunks "
+                    "with different example values."
+                )
+
+            additional_kwargs = merge_dicts(
+                self.additional_kwargs, other.additional_kwargs
+            )
+            raw_tool_calls = additional_kwargs.get("tool_calls", [])
+            tool_calls = [
+                ToolCall(**tool_call)
+                for tool_call in parse_tool_calls(
```

Comment:
hm so would this only work for openai

Reply:
i think we may want to drop this and rely on each model handling accumulating tool calls (in our format) itself, and here we just combine tool_calls lists

Reply:
Updated this. For clarity, as of now when chunks are accumulated, `additional_kwargs` gets updated via `utils._merge.merge_dicts` (this may have OpenAI-specific logic in its [references to "index"](https://github.com/langchain-ai/langchain/blob/83f62fdacfa7930c975e2c4aea5733b17e8d6bde/libs/core/langchain_core/utils/_merge.py#L38), not sure if this is standard).

Tool call lists are appended, so after accumulating a few message chunks you do get things like
```
[ToolCall(name='', args={}),
 ToolCall(name='', args={}),
 ToolCall(name='', args={}),
 ToolCall(name='', args={})]
```
in `.tool_calls` in the resulting chunk (whereas the additional_kwargs looks reasonable).

During streaming, output parsers continue to parse the accumulated additional_kwargs. For `ToolCallsMessage` they can read off the parsed .tool_calls.